### ISSUE TITLE
SA1307 considers '_' as a UpperCase letter

### ISSUE THREAD

**Issue by tobias-wiese** [DotNetAnalyzers/StyleCopAnalyzers#3559]
SA1307 states "The name of a public or internal field in C# does not begin with an upper-case letter.".

However, this code is not reported:

```cs
public partial class SomeClass
    {
        public static Instance _instance;
        //...
    }
```

AFAIK the `_` is convenionally used for private fields, but if it is used for public fields (e.g. by mistake) it should be either reported by an own check, or by 1307 in this case.

The reason is this part of the code in https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1307AccessibleFieldsMustBeginWithUpperCaseLetter.cs :
```cs
if (!string.IsNullOrEmpty(name) 
    && char.IsLower(name[0])
    && !NamedTypeHelpers.IsContainedInNativeMethodsClass(declaration))
{
    context.ReportDiagnostic(Diagnostic.Create(Descriptor, declarator.Identifier.GetLocation(), name));
}
```
Here, `char.isLower(name[0])` will return `False` for names beginning with `_`.

I would expect that in such cases either the first char after the `_` is checked, or the `_` is reported at first.

**Comment by bjornhellander** [comment]
Hi!

The documentation for SA1307 also says this:
_A violation of this rule occurs when the name of a public or internal field begins with a lower-case letter_

Leading underscore is handled in SA1309 instead and it triggers for me. Have you turned this one off?

For situations where SA1309 is turned off, I can also see a benefit in changing SA1307 the way you have described. I am however unsure if that is desirable, since there are so much history related to these rules. @sharwell?

**Comment by tobias-wiese** [comment]
Yes, we had this off since it the creates findings for private, public, internal and protected fields. We would like to limit this to visible fields (public/internal) in this case wrt the C# naming conventions (https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#naming-conventions). WDYT?

**Comment by tobias-wiese** [comment]
@bjornhellander any news on this? :)

**Comment by bjornhellander** [comment]
Up to @sharwell to decide
